Dockerfile

```diff
@@ -0,0 +1,49 @@
+FROM python:3.12 AS python-base
+
+RUN apt-get update -y
+RUN passwd --delete root
+RUN adduser cotainr --disabled-password --gecos ""
+
+RUN pip install pip --upgrade
+COPY test-requirements.txt /tmp/requirements/test-requirements.txt
+COPY docs-requirements.txt /tmp/requirements/docs-requirements.txt
+RUN pip install -r /tmp/requirements/test-requirements.txt
+RUN pip install -r /tmp/requirements/docs-requirements.txt
+
+WORKDIR /home/cotainr
+COPY . /home/cotainr/
+USER root
+ENV PATH="/home/cotainr/bin:$PATH"
+
+FROM python-base AS go-base
+
+RUN apt-get update
+RUN apt-get install -y build-essential \
+    uuid-dev \
+    libgpgme-dev \
+    squashfs-tools \
+    libseccomp-dev \
+    wget \
+    pkg-config \
+    git \
+    cryptsetup-bin
+
+RUN export VERSION=1.17.6 OS=linux ARCH=amd64 && \
+    wget https://dl.google.com/go/go$VERSION.$OS-$ARCH.tar.gz && \
+    tar -C /usr/local -xzvf go$VERSION.$OS-$ARCH.tar.gz && \
+    rm go$VERSION.$OS-$ARCH.tar.gz
+
+ENV GOPATH="/home/cotainr/go"
+ENV PATH="/usr/local/go/bin:$PATH:/home/cotainr/go/bin"
+
+FROM go-base AS singularity
+
+# adjust this as necessary
+RUN export VERSION=3.8.7 && \
+    wget https://github.com/hpcng/singularity/releases/download/v${VERSION}/singularity-${VERSION}.tar.gz && \
+    tar -xzf singularity-${VERSION}.tar.gz
+WORKDIR /home/cotainr/singularity-3.8.7
+RUN ./mconfig
+RUN make -C ./builddir
+RUN make -C ./builddir install
+WORKDIR /home/cotainr
```

Comment:
Wouldn't it be much easier to use a Containerized Singularity base image? https://quay.io/repository/singularity/singularity

Reply:
I think it is a good suggestion, and I have attempted this

```Dockerfile
FROM quay.io/singularity/singularity:v4.1.0 AS singularity-base

# glibc compatibility libraries
RUN apk add gcompat

# Install python/pip
ENV PYTHONUNBUFFERED=1
RUN apk add --update --no-cache python3 && ln -sf python3 /usr/bin/python
RUN python3 -m ensurepip
RUN pip3 install --no-cache --upgrade pip setuptools

# Install project requirements
COPY test-requirements.txt /tmp/requirements/test-requirements.txt
COPY docs-requirements.txt /tmp/requirements/docs-requirements.txt
RUN pip install -r /tmp/requirements/test-requirements.txt
RUN pip install -r /tmp/requirements/docs-requirements.txt

# Initialize environment
WORKDIR /home/cotainr
COPY . /home/cotainr/
ENV PATH="/home/cotainr/bin:$PATH"
ENTRYPOINT [ ]
```
However, it seems like Anaconda's python is linked to glibc, which is not included in the alpine based base images you linked, and the gcompat library does not seem to fix this issue.
As a result the test at `pack/test_conda_install.py::Test_DisplayMiniforgeLicenseForAcceptance::test_miniforge_still_showing_license` fails. 

I think we should build the docker with, if not the same OS as the CI, then at least Debian based one, like python:3.12.